crates/host/src/module.rs

```diff
@@ -41,53 +139,109 @@ impl SerializedModuleCache {
 }
 
 #[derive(Default)]
-pub struct ModuleCache(HashMap<[u8; 32], Arc<Module>>, bool, AtomicUsize);
+pub struct ModuleCache {
+    plru: MicroCache,
+    key_map: PlruKeyMap,
+    cache: HashMap<CacheKey, Arc<Module>>,
+    leak_buster: AtomicUsize,
+}
 
 pub static MODULE_CACHE: Lazy<RwLock<ModuleCache>> =
     Lazy::new(|| RwLock::new(ModuleCache::default()));
 
 impl ModuleCache {
+    pub fn reset_leak_buster(&mut self) {
+        self.leak_buster
+            .store(0, std::sync::atomic::Ordering::SeqCst);
+    }
+
+    pub fn should_bust_leak(&mut self) -> bool {
+        if self
+            .leak_buster
+            .fetch_add(1, std::sync::atomic::Ordering::SeqCst)
+            > 100
+        {
+            self.reset_leak_buster();
+            true
+        } else {
+            false
+        }
+    }
+
     fn get_with_build_cache(
         &mut self,
-        key: [u8; 32],
+        key: CacheKey,
         wasm: &[u8],
     ) -> Result<Arc<Module>, WasmError> {
         let module = SERIALIZED_MODULE_CACHE.write().get(key, wasm)?;
-        let arc = Arc::new(module);
-        self.0.insert(key, Arc::clone(&arc));
-        Ok(arc)
-    }
-
-    pub fn reset_counter(&mut self) {
-        self.2.store(0, std::sync::atomic::Ordering::SeqCst);
-    }
-
-    pub fn get(&mut self, key: [u8; 32], wasm: &[u8]) -> Result<Arc<Module>, WasmError> {
-        let count = self.2.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
-        if count > 100 {
-            let current = self.0.remove(&key);
-            match current {
-                Some(current) => match Arc::try_unwrap(current) {
-                    Ok(m) => {
-                        std::mem::drop(m);
-                        self.2.store(0, std::sync::atomic::Ordering::SeqCst);
-                        self.get_with_build_cache(key, wasm)
-                    }
-                    Err(current) => {
-                        self.0.insert(key, current.clone());
-                        Ok(current)
-                    }
-                },
-                None => {
-                    self.2.store(0, std::sync::atomic::Ordering::SeqCst);
-                    self.get_with_build_cache(key, wasm)
-                }
-            }
+        Ok(self.put_item(key, module))
+    }
+
+    pub fn get(&mut self, key: CacheKey, wasm: &[u8]) -> Result<Arc<Module>, WasmError> {
+        match if self.should_bust_leak() {
+            self.remove_item(&key)
         } else {
-            match self.0.get(&key) {
-                Some(module) => Ok(Arc::clone(module)),
-                None => self.get_with_build_cache(key, wasm),
-            }
+            self.get_item(&key)
+        } {
+            Some(module) => Ok(module),
+            None => self.get_with_build_cache(key, wasm),
```

Comment:
So the reason we did the try_unwrap was because this doesn't check if any of the `Arc<Module>` are still around before trying to deserialize a new one. Wasmer panics if you deserialize a module while the same module hasn't been dropped.

Reply:
I can't reproduce this so I guess we wait till we encounter it in the wild again